Source/Core/Resource.js

```diff
@@ -0,0 +1,534 @@
+
+        //>>includeStart('debug', pragmas.debug);
+        Check.typeOf.string('options.url', options.url);
+        //>>includeEnd('debug');
+
+        this._url = '';
+        this._templateValues = encodeValues(defaultValue(options.templateValues, {}));
+        this._queryParameters = defaultValue(options.queryParameters, {});
+        this._isDirectory = defaultValue(options.isDirectory, false);
+        this.url = options.url;
+
+        /**
+         * Additional HTTP headers that will be sent with the request.
+         *
+         * @type {Object}
+         */
+        this.headers = defined(options.headers) ? clone(options.headers) : {};
+
+        /**
+         * A Request object that will be used. Intended for internal use only.
+         *
+         * @type {Request}
+         */
+        this.request = options.request;
+
+        /**
+         * The type of response expected from the request.
+         *
+         * @type {String}
+         */
+        this.responseType = options.responseType;
+
+        /**
+         * The method to use for the request.
+         *
+         * @type {String}
+         */
+        this.method = defaultValue(options.method, 'GET');
+
+        /**
+         * Data to be sent with the request.
+         *
+         * @type {Object}
+         */
+        this.data = options.data;
+
+        /**
+         * Overrides the MIME type returned by the server.
+         *
+         * @type {String}
+         */
+        this.overrideMimeType = options.overrideMimeType;
+
+        /**
+         * A proxy to be used when loading the resource.
+         *
+         * @type {DefaultProxy}
+         */
+        this.proxy = options.proxy;
+
+        /**
+         * Function to call when a request for this resource fails. If it returns true, the request will be retried.
+         *
+         * @type {Function}
+         */
+        this.retryCallback = options.retryCallback;
+
+        /**
+         * The number of times the retryCallback should be called before giving up.
+         *
+         * @type {Number}
+         */
+        this.retryAttempts = defaultValue(options.retryAttempts, 0);
+        this._retryCount = 0;
+    }
+
+    /**
+     * A helper function to create a resource depending on whether we have a String or a Resource
+     *
+     * @param {Resource|String} resource A Resource or a String to use when creating a new Resource.
+     * @param {Object{ options If resource is a String, these are the options passed to the Resource constructor. It is ignored otherwise.
+     *
+     * @returns {Resource} If resource is a String, a Resource constructed with the url and options. Otherwise the resource parameter is returned.
+     *
+     * @private
+     */
+    Resource.createIfNeeded = function(resource, options) {
+        if (!defined(resource)) {
+            return;
+        }
+
+        if (typeof resource === 'string') {
+            var args = defined(options) ? clone(options) : {};
```

Comment:
Is this clone needed?  it looks like everywhere we use `Resource.createIfNeeded` we create a throwaway anonymous object for options.